core/cap-0046-10.md

```diff
@@ -0,0 +1,229 @@
+
+A model with higher order dependencies also risk the worst-case costs significantly outweighing the average, and any small deviation in the input resulting in significant over or underestimation of the costs. This goes against the design goals. 
+
+
+### Host vs WASM vm
+This metering framework is generic and does not differentiate between the host and the WASM vm. Both the host and the vm are treated as components and blocks defined in the "specification" section and subject to the same metering procedures. 
+
+Our current choice of the WASM virtual machine implementation is Wasmi, which is a lightweight interpreter of the wasm standard, written in the same language (Rust) as the host. Wasmi runs an inner interpreter loop that executes a single wasm instruction on each loop. Thus every wasm instruction logic fits the requirements of a component. `WasmInsnExecT0~4` in `ContractCostType` are designated for the wasm instructions (instead of having one type designated to each of the 100+ wasm instructions, we group them into tiers 0~4 where each tier of wasm instructions costs relatively the same amount of cpu insns).
+
+We maintain a fork of Wasmi with metering added. This makes Wasmi is a tamed "wild component". 
+
+(Note this does not mean we are tied to a particular wasm implementation, it's just an example. If we decide to switch to a different interpreter or JIT in the future, we will be able to apply the same procedure to derive a new set of metering components.)
+
+
+### Relation to cap-0046-07 (fee model)
+
+[CAP-0046-07](./cap-0046-07.md) proposed a fee model for smart contracts taking into account ledger access, storage and computation (or "gas"). This CAP details the computation aspect. However, this proposal identifies cpu and memory as separate aspects of the compute cost that needs to be budgeted separately. This difference needs to be resolved before this CAP finalize, i.e., either expand gas network settings in 07 or consolidate the `cpu_insns` and `mem_bytes` into a single "gas" parameter in here. 
+
+### Cost estimation
+This proposal relies on the "preflight" mechanism to provide users with cost estimation of a transaction. The total costs for each resource type as well as inputs to each individual cost type will be returned from the preflight simulation. These costs, however can only serve as guidance to the actual cost, since the ledger snapshot used for preflight may be outdated. Thus it is not guaranteed that a transaction staying below the budget during preflight will not exceed it during the actual run. 
+
+## Parameters Upgrade
+Both the budget and metering parameters are stored on the ledger as `ConfigLedgerEntry` and their upgrade and validation process have been discussed in [CAP-0046-09](./cap-0046-09.md). In general, the parameters can be upgraded with or without a protocol upgrade. 
```

Comment:
I think you mean protocol *version* upgrade here. IIUC all the upgrades are protocol upgrades (such as changing max tx set size or any parameter value here), while the version upgrades allow for some bigger changes that (basically any XDR changes).

Reply:
Thanks. I've reworded.